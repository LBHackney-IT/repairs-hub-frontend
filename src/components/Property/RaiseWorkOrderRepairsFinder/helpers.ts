import { parseString } from 'xml2js'
import { Priority } from '@/root/src/models/priority'

export const isOutOfHoursGas = (
  contractorReference: string,
  tradeCode: string
) => {
  const gasBreakdownContractorReference = 'H04'
  const oohTradeCode = 'OO'

  if (contractorReference != gasBreakdownContractorReference) return false // contractor must be "H04"
  return tradeCode == oohTradeCode
}

export const getPriorityObjectByDescription = (
  description: string,
  priorities: Priority[]
) => {
  return priorities.find((priority) => priority.description === description)
}

export const getPriorityObjectByCode = (
  code: number,
  priorities: Priority[]
) => {
  return priorities.find((priority) => priority.priorityCode == code)
}

export interface RepairsFinderExtractedData {
  sorCode: string
  priority: string
  quantity: string
  tradeCode: string
  comments: string
  contractorReference: string
}

export const extractXmlData = async (
  xmlContent: string
): Promise<RepairsFinderExtractedData> => {
  const result = await handleParseXML(xmlContent)

  if (!result.success) {
    return null
  }

  let {
    PRIORITY: [priority],
    QUANTITY: [quantity],
    SOR_CODE: [sorCode],
    SOR_COMMENTS: [comments],
    SOR_CLASS: [tradeCode],
  } = result.result.RF_INFO.SOR[0]

  let contractorReference = result.result.RF_INFO.WORK_PROGRAMME[0]

  // temp override
  sorCode = '20060020'
  priority = 'Normal'
  quantity = '1'
  tradeCode = 'PL'
  comments = 'Sink top is loose - sadfsdf'
  contractorReference = 'H01'

  return {
    sorCode,
    priority,
    quantity,
    tradeCode,
    comments,
    contractorReference,
  }
}

interface RepairsFinderXmlContent {
  RF_INFO: {
    RESULT: string[]
    PROPERTY: string[]
    WORK_PROGRAMME: string[]
    CAUSED_BY: string[]
    NOTIFIED_DEFECT: string[]
    DEFECT: Array<{
      DEFECT_CODE: string[]
      DEFECT_LOC_CODE: string[]
      DEFECT_COMMENTS: string[]
      DEFECT_PRIORITY: string[]
      DEFECT_QUANTITY: string[]
    }>
    SOR: Array<{
      SOR_CODE: string[]
      PRIORITY: string[]
      QUANTITY: string[]
      SOR_LOC_CODE: string[]
      SOR_COMMENTS: string[]
      SOR_CLASS: string[]
    }>
  }
}

const handleParseXML = (xml: string) =>
  new Promise<{
    success: boolean
    error: Error
    result: RepairsFinderXmlContent
  }>((resolve) => {
    parseString(xml, (err, result) => {
      if (err) {
        resolve({ success: false, error: err, result: null })
        return
      }

      resolve({ success: true, error: null, result })
    })
  })
