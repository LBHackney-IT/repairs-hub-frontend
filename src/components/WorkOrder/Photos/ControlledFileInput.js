import { useRef } from 'react'

import ErrorMessage from '../../Errors/ErrorMessage'
import PhotoUploadPreview from './PhotoUploadPreview'
import classNames from 'classnames'
import useUpdateFileInput from './hooks/useUpdateFileInput'

const ControlledFileInput = ({
  files,
  setFiles,
  validationError,
  isLoading,
  register,
  label = 'Photos',
  labelSize = 's',
  disabled = false,
  showAsOptional = false,
}) => {
  const inputRef = useRef()

  // extracted to enable mocking
  useUpdateFileInput(files, inputRef)

  return (
    <div>
      <legend
        className={`govuk-fieldset__legend govuk-fieldset__legend--${labelSize}`}
      >
        {label} {showAsOptional && '(optional) '}
      </legend>

      <span id={`${'photos'}-hint`} className="govuk-hint lbh-hint">
        Select all the photos you want to add (up to 10 photos)
      </span>

      {validationError && <ErrorMessage label={validationError} />}
      <input
        disabled={disabled}
        ref={inputRef}
        name="fileUpload"
        id="fileUpload"
        data-testid="fileUploadInput"
        className={classNames('govuk-file-upload custom-file-input', {
          'govuk-form-group--error': validationError,
        })}
        type="file"
        multiple
        accept=".jpg, .jpeg, .png"
        onInput={(e) => {
          setFiles(Object.values(e.target.files))
        }}
        style={{
          marginTop: '10px',
        }}
        {...register}
      />

      <>
        {files.length > 0 && (
          <PhotoUploadPreview
            files={files}
            disabled={isLoading}
            setFiles={setFiles}
          />
        )}
      </>
    </div>
  )
}

export default ControlledFileInput
