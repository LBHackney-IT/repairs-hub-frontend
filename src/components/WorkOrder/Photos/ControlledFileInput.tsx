import { Dispatch, SetStateAction, useEffect, useRef, useState } from 'react'

import ErrorMessage from '../../Errors/ErrorMessage'
import PhotoUploadPreview from './PhotoUploadPreview'
import classNames from 'classnames'
import useUpdateFileInput from './hooks/useUpdateFileInput'
import {
  cachedFileExists,
  getCachedFile,
  setCachedFile,
} from './hooks/uploadFiles/cacheFile'
import SpinnerWithLabel from '../../SpinnerWithLabel'
import compressFile from './hooks/uploadFiles/compressFile'

interface Props {
  files: File[]
  setFiles: Dispatch<SetStateAction<File[]>>
  validationError: string
  isLoading: boolean
  register: any
  label: string
  hint: string
  disabled?: boolean
  testId: string
}

const ControlledFileInput = (props: Props) => {
  const {
    files,
    setFiles,
    validationError,
    isLoading,
    register,
    label,
    hint,
    disabled = false,
    testId,
  } = props

  const inputRef = useRef<HTMLInputElement | null>(null)
  const [isCompressing, setIsCompressing] = useState(false)
  const [compressedCount, setCompressedCount] = useState(0)
  const [totalFilesToCompress, setTotalFilesToCompress] = useState(0)
  const [previewFiles, setPreviewFiles] = useState<File[]>([])

  useUpdateFileInput(inputRef, files)

  useEffect(() => {
    const neededFiles = files.filter(
      (file) => !previewFiles.some((pf) => pf.name === file.name)
    )
    setPreviewFiles((prev) => [...prev, ...neededFiles])
  }, [files, previewFiles])

  const handleInput = async (e: React.ChangeEvent<HTMLInputElement>) => {
    const selectedFiles = Array.from(e.target.files || [])

    setFiles([])
    setPreviewFiles([])
    setIsCompressing(true)
    setTotalFilesToCompress(selectedFiles.length)
    setCompressedCount(0)

    try {
      const stableFiles: File[] = []

      // Process files one-by-one to avoid memory spikes
      // Read each file into memory and create a stable copy away from the OS
      for (const file of selectedFiles) {
        const buffer = await file.arrayBuffer()
        const stableFile = new File([buffer], file.name, { type: file.type })
        stableFiles.push(stableFile)
      }
      setFiles(stableFiles)

      for (const file of stableFiles) {
        if (await cachedFileExists(file)) {
          const cached = await getCachedFile(file)
          if (cached) {
            console.log('File already cached, skipping compression:', file.name)
            setPreviewFiles((prev) => [...prev, cached])
            setCompressedCount((prevCount) => prevCount + 1)
            continue
          }
        }
        try {
          const compressed = await compressFile(file)
          setCachedFile(compressed)
          setPreviewFiles((prev) => [...prev, compressed])
        } catch (error) {
          console.error('Error compressing file:', error)
          setCachedFile(file)
          setPreviewFiles((prev) => [...prev, file])
        }
        setCompressedCount((prevCount) => prevCount + 1)
      }
    } catch (err) {
      console.error('Error processing files:', err)
    } finally {
      setIsCompressing(false)
    }
  }

  return (
    <div>
      <legend
        id="fileUploadLegend"
        className={`govuk-fieldset__legend govuk-fieldset__legend--s`}
      >
        {label}
      </legend>

      <span id="photos-hint" className="govuk-hint lbh-hint">
        {hint}
      </span>

      {validationError && <ErrorMessage label={validationError} />}

      <input
        disabled={disabled}
        ref={inputRef}
        name="fileUpload"
        id="fileUpload"
        aria-labelledby="fileUploadLegend"
        data-testid={testId}
        className={classNames('govuk-file-upload custom-file-input', {
          'govuk-form-group--error': validationError,
        })}
        type="file"
        multiple
        accept=".jpg, .jpeg, .png"
        onInput={handleInput}
        style={{
          marginTop: '10px',
        }}
        {...register}
      />

      <>
        {previewFiles.length > 0 && (
          <>
            <PhotoUploadPreview
              files={previewFiles}
              disabled={isLoading}
              setFiles={(newPreviewFiles: File[]) => {
                setPreviewFiles(newPreviewFiles)
                const keepNames = new Set(newPreviewFiles.map((f) => f.name))
                const newFormFiles = files.filter((f) => keepNames.has(f.name))
                setFiles(newFormFiles)
              }}
            />
          </>
        )}
        {isCompressing && (
          <div style={{ display: 'flex', alignItems: 'center', gap: 12 }}>
            <SpinnerWithLabel
              label={`Caching photos... (${compressedCount} of ${totalFilesToCompress})`}
            />
          </div>
        )}
      </>
    </div>
  )
}

export default ControlledFileInput
