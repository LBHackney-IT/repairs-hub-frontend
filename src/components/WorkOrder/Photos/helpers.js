export const mockFile = (name) => {
  const dataBase64 = 'VEhJUyBJUyBUSEUgQU5TV0VSCg=='

  const arrayBuffer = Uint8Array.from(window.atob(dataBase64), (c) =>
    c.charCodeAt(0)
  )

  const blob = new Blob([arrayBuffer], { type: 'text/html' })
  const file = new File([blob], name, { type: 'text/html' })

  return file
}
