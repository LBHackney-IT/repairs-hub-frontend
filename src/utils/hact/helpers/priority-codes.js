import {
  IMMEDIATE_PRIORITY_CODE,
  EMERGENCY_PRIORITY_CODE,
  URGENT_PRIORITY_CODE,
  NORMAL_PRIORITY_CODE,
} from '../../helpers/priorities'

export const priorityCodeCompletionTimes = {
  [IMMEDIATE_PRIORITY_CODE]: {
    numberOfHours: 2,
    numberOfDays: 0,
  },
  [EMERGENCY_PRIORITY_CODE]: {
    numberOfHours: 24,
    numberOfDays: 1,
  },
  [URGENT_PRIORITY_CODE]: {
    numberOfHours: 120,
    numberOfDays: 5,
  },
  [NORMAL_PRIORITY_CODE]: {
    numberOfHours: 504,
    numberOfDays: 21,
  },
  // inspection ?
  // 5: {
  //   numberOfHours: 8670,
  //   numberOfDays: 365,
  // },
}
