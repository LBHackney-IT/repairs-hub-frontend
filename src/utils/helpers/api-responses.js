import HttpStatus from 'http-status-codes'

export const isSpendLimitReachedResponse = (response) =>
  response?.status === HttpStatus.UNAUTHORIZED &&
  response?.data?.message?.toLowerCase().match('limit')
