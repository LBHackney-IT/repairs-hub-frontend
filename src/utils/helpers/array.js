export const uniqueArrayValues = (array) => {
  const set = new Set(array)
  return [...set]
}

export const sortArrayByDate = (array, dateAttribute) => {
  return array.sort(function (a, b) {
    return new Date(b[dateAttribute]) - new Date(a[dateAttribute])
  })
}
