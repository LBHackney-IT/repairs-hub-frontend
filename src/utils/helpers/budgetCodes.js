export const formatBudgetCodeForOption = (
  budgetCode,
  contractorIsPurdy,
  separator = ' - '
) => {
  const displayFields = [
    budgetCode?.externalCostCode,
    budgetCode?.corporateSubjectiveCode,
    budgetCode?.descriptionOfWorks,
  ]

  if (!contractorIsPurdy) displayFields.push(budgetCode?.contractorList)

  return displayFields.filter((x) => x).join(separator)
}

export const formatBudgetCode = (budgetCode) => {
  const displayFields = [
    budgetCode?.externalCostCode,
    [budgetCode?.corporateSubjectiveCode, budgetCode?.descriptionOfWorks]
      .filter((x) => x)
      .join(' '),
  ]

  return displayFields.filter((x) => x).join(' - ')
}
