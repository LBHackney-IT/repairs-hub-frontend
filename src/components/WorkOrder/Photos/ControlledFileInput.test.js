import { render } from '@testing-library/react'
import ControlledFileInput from './ControlledFileInput'
import { mockFile } from './helpers'

jest.mock(
  './hooks/useUpdateFileInput',
  jest.fn((files, inputRef) => {
    inputRef.current.value = ''
  })
)

describe('ControlledFileInput component', () => {
  beforeAll(() => {
    global.URL = {
      createObjectURL: jest.fn(() => 'mockResponseUrl'),
    }
  })

  it('renders component when no files selected', () => {
    const { asFragment } = render(
      <ControlledFileInput
        files={[]}
        setFiles={() => {}}
        validationError={null}
        isLoading={false}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders shows file previews', () => {
    const file = mockFile('test.png')

    const { asFragment } = render(
      <ControlledFileInput
        files={[file]}
        setFiles={() => {}}
        validationError={null}
        isLoading={false}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders shows validation error', () => {
    const file = mockFile('test.png')

    const { asFragment } = render(
      <ControlledFileInput
        files={[file]}
        setFiles={() => {}}
        validationError={'Some validation error'}
        isLoading={false}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
