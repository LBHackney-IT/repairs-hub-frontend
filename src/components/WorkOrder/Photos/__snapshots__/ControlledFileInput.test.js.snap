// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlledFileInput component renders component when no files selected 1`] = `
<DocumentFragment>
  <div>
    <legend
      class="govuk-fieldset__legend govuk-fieldset__legend--s"
      id="fileUploadLegend"
    />
    <span
      class="govuk-hint lbh-hint"
      id="photos-hint"
    />
    <input
      accept=".jpg, .jpeg, .png"
      aria-labelledby="fileUploadLegend"
      class="govuk-file-upload custom-file-input"
      id="fileUpload"
      multiple=""
      name="fileUpload"
      style="margin-top: 10px;"
      type="file"
    />
  </div>
</DocumentFragment>
`;

exports[`ControlledFileInput component renders shows file previews 1`] = `
<DocumentFragment>
  <div>
    <legend
      class="govuk-fieldset__legend govuk-fieldset__legend--s"
      id="fileUploadLegend"
    />
    <span
      class="govuk-hint lbh-hint"
      id="photos-hint"
    />
    <input
      accept=".jpg, .jpeg, .png"
      aria-labelledby="fileUploadLegend"
      class="govuk-file-upload custom-file-input"
      id="fileUpload"
      multiple=""
      name="fileUpload"
      style="margin-top: 10px;"
      type="file"
    />
    <div
      class="photoUploadPreview"
    >
      <div
        class="photoUploadPreview-container"
      >
        <div
          class="photoUploadPreview-imageContainer"
        >
          <img
            alt="Preview Uploaded Image"
            class="photoUploadPreview-image"
            id="file-preview"
            src="mockResponseUrl"
          />
        </div>
        <button
          class="photoUploadPreview-removeButton"
          type="button"
        >
          Remove
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ControlledFileInput component renders shows validation error 1`] = `
<DocumentFragment>
  <div>
    <legend
      class="govuk-fieldset__legend govuk-fieldset__legend--s"
      id="fileUploadLegend"
    />
    <span
      class="govuk-hint lbh-hint"
      id="photos-hint"
    />
    <span
      class="govuk-error-message lbh-error-message"
      data-error-id=""
      data-testid="error-message"
    >
      <span
        class="govuk-visually-hidden"
      >
        Error:
      </span>
       Some validation error
    </span>
    <input
      accept=".jpg, .jpeg, .png"
      aria-labelledby="fileUploadLegend"
      class="govuk-file-upload custom-file-input govuk-form-group--error"
      id="fileUpload"
      multiple=""
      name="fileUpload"
      style="margin-top: 10px;"
      type="file"
    />
    <div
      class="photoUploadPreview"
    >
      <div
        class="photoUploadPreview-container"
      >
        <div
          class="photoUploadPreview-imageContainer"
        >
          <img
            alt="Preview Uploaded Image"
            class="photoUploadPreview-image"
            id="file-preview"
            src="mockResponseUrl"
          />
        </div>
        <button
          class="photoUploadPreview-removeButton"
          type="button"
        >
          Remove
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
